File: day02/puzzle1.cpp
#include <fstream>
#include <iostream>
#include <vector>
#include <string>
#include <set>
#include <unordered_set>
#include <map>
#include <unordered_map>

using namespace std;

int main() {
  ifstream in("input");
  int good = 0;
  while (in) {
    int lo, hi;
    char letter;
    char trash;
    string pass;
    in >> lo;
    in >> trash;
    in >> hi;
    in >> letter;
    in >> trash;
    in >> pass;
    int count = 0;
    for (int i = 0; i < pass.length(); i++) {
      if (pass[i] == letter) count++;
    }
    if (count >= lo && count <= hi) good++;
  }
  cout << good << endl;
}

File: day02/puzzle1.js
const fs = require('fs');

const IGNORE_EMPTY_LINES = true;

const data = fs.readFileSync('input', 'UTF-8');
let lines = data.split(/\r?\n/);

if (IGNORE_EMPTY_LINES) {
  lines = lines.filter(line => line !== "");
}

const valid =
  lines
    .map(line => {
      console.log(line);
      const [range, letterColon, pass] = line.split(' ');
      const [lo, hi] = range.split('-').map(n => parseInt(n, 10));
      const letter = letterColon[0];
      return {hi, lo, letter, pass};
    })
    .reduce((acc, {lo, hi, letter, pass}) => {
      const count = pass.split(letter).length-1;
      return acc + Number(lo <= count && count <= hi);
    }, 0);

console.log(valid);
